File: app/services/serie_service.py
"""Serie service that uses MongoDB when available, otherwise falls back to in-memory storage.
This ports the logic from the original Node.js implementation (uploading thumbnails, SNS topic management,
and updating MongoDB collections).
"""
from uuid import uuid4
from app.utils.mongodb import connect_to_database
from app.utils.s3 import upload_via_cloudfront, delete_via_cloudfront
from app.utils.sns import create_topic, delete_topic, subscribe_to_serie, unsubscribe_from_topic

_SERIES = {}
_SUBSCRIPTIONS = {}


def _db():
    return connect_to_database()


def create_serie(data, user_id=None, id_token=None, file=None):
    db = _db()
    # if we have a real DB, run Mongo logic similar to Node
    if db:
        series_col = db.collection("series")
        image_url = ""
        if file:
            unique_name = f"{uuid4()}_{getattr(file, 'filename', 'file')}"
            buffer = getattr(file, 'read', lambda: None)()
            mimetype = getattr(file, 'mimetype', None) or getattr(file, 'content_type', None)
            image_url = upload_via_cloudfront(id_token, buffer, unique_name, mimetype, f"files/user-{user_id}/thumbnail")

        new_serie = {
            **data,
            "serie_thumbnail": image_url,
            "isPublish": data.get("isPublish", False),
            "serie_user": user_id,
            "serie_lessons": data.get("serie_lessons", []),
            "createdAt": None,
            "updatedAt": None,
            "serie_subcribe_num": 0,
        }
        result = series_col.insert_one(new_serie)
        inserted_id = str(result.inserted_id)
        topic_arn = create_topic(f"serie_{inserted_id}")
        series_col.update_one({"_id": result.inserted_id}, {"$set": {"serie_sns": topic_arn}})
        return {"_id": inserted_id, **new_serie, "serie_sns": topic_arn}

    # fallback: in-memory
    serie_id = str(uuid4())
    serie = {"_id": serie_id, **data, "created_by": user_id}
    _SERIES[serie_id] = serie
    return serie


def get_all_series(query=None):
    db = _db()
    if db:
        serie_col = db.collection("series")
        q = dict(query) if query else {}
        return list(serie_col.find(q))
    return list(_SERIES.values())


def get_serie_by_id(serie_id):
    db = _db()
    if db:
        try:
            from bson import ObjectId

            if not ObjectId.is_valid(serie_id):
                return None
            serie_col = db.collection("series")
            return serie_col.find_one({"_id": ObjectId(serie_id)})
        except Exception:
            return None
    return _SERIES.get(serie_id)


def get_all_series_by_user(user_id):
    db = _db()
    if db:
        serie_col = db.collection("series")
        return list(serie_col.find({"serie_user": user_id}))
    return [s for s in _SERIES.values() if s.get("created_by") == user_id]


def search_series_by_title(keyword):
    db = _db()
    if db:
        serie_col = db.collection("series")
        # Use text search if index exists, otherwise simple regex
        try:
            return list(serie_col.find({"$text": {"$search": keyword}, "isPublish": True}))
        except Exception:
            return list(serie_col.find({"serie_title": {"$regex": keyword, "$options": "i"}, "isPublish": True}))
    keyword_lower = keyword.lower()
    return [s for s in _SERIES.values() if keyword_lower in str(s.get("serie_title", "")).lower()]


def get_series_subscribed_by_user(user_id):
    db = _db()
    if db:
        user_col = db.collection("users")
        serie_col = db.collection("series")
        user = user_col.find_one({"_id": user_id}, {"serie_subcribe": 1})
        if not user or not user.get("serie_subcribe"):
            return []
        serie_ids = user.get("serie_subcribe")
        # Attempt to convert to ObjectId when necessary
        from bson import ObjectId

        obj_ids = []
        for sid in serie_ids:
            try:
                obj_ids.append(ObjectId(sid))
            except Exception:
                pass
        return list(serie_col.find({"_id": {"$in": obj_ids}}))
    # in-memory fallback
    result = []
    for sid, subs in _SUBSCRIPTIONS.items():
        if user_id in subs:
            serie = _SERIES.get(sid)
            if serie:
                result.append(serie)
    return result


def update_serie(serie_id, data, user_id=None, id_token=None, file=None):
    db = _db()
    if db:
        serie_col = db.collection("series")
        # coerce isPublish string
        if isinstance(data.get("isPublish"), str):
            data["isPublish"] = data["isPublish"].lower() == "true"
        if file:
            # delete old and upload new
            current = serie_col.find_one({"_id": serie_col.ObjectId(serie_id)})
            if current and current.get("serie_thumbnail"):
                delete_via_cloudfront(current.get("serie_thumbnail"))
            unique_name = f"{uuid4()}_{getattr(file,'filename','file')}"
            buffer = getattr(file, 'read', lambda: None)()
            mimetype = getattr(file, 'mimetype', None) or getattr(file, 'content_type', None)
            new_url = upload_via_cloudfront(id_token, buffer, unique_name, mimetype, f"files/user-{user_id}/thumbnail")
            data["serie_thumbnail"] = new_url
        data["updatedAt"] = None
        res = serie_col.update_one({"_id": serie_col.ObjectId(serie_id)}, {"$set": data})
        if res.matched_count == 0:
            return None
        return serie_col.find_one({"_id": serie_col.ObjectId(serie_id)})
    existing = _SERIES.get(serie_id)
    if not existing:
        return None
    if file:
        # no-op for in-memory
        data["serie_thumbnail"] = f"uploaded://{getattr(file,'filename','file')}"
    existing.update(data)
    _SERIES[serie_id] = existing
    return existing


def subscribe_serie(serie_id, user_id, user_email):
    db = _db()
    if db:
        serie_col = db.collection("series")
        user_col = db.collection("users")
        serie = serie_col.find_one({"_id": serie_col.ObjectId(serie_id)})
        if not serie or not serie.get("serie_sns"):
            raise ValueError("Serie not found")
        user = user_col.find_one({"_id": user_id})
        if not user:
            raise ValueError("User not found")
        if user.get("serie_subcribe") and serie_id in user.get("serie_subcribe"):
            return {"message": "Bạn đã đăng ký series này rồi.", "alreadySubscribed": True}
        subscribe_to_serie(serie.get("serie_sns"), user_email)
        user_col.update_one({"_id": user_id}, {"$addToSet": {"serie_subcribe": serie_id}, "$set": {"updatedAt": None}})
        serie_col.update_one({"_id": serie_col.ObjectId(serie_id)}, {"$inc": {"serie_subcribe_num": 1}, "$set": {"updatedAt": None}})
        return {"message": "Subscribed"}
    subs = _SUBSCRIPTIONS.setdefault(serie_id, set())
    if user_id in subs:
        return {"message": "Bạn đã đăng ký series này rồi.", "alreadySubscribed": True}
    subs.add(user_id)
    return {"message": "Đăng ký nhận thông báo thành công.", "result": {"serieId": serie_id, "userId": user_id}}


def unsubscribe_serie(serie_id, user_id, user_email):
    db = _db()
    if db:
        serie_col = db.collection("series")
        user_col = db.collection("users")
        serie = serie_col.find_one({"_id": serie_col.ObjectId(serie_id)})
        if not serie or not serie.get("serie_sns"):
            raise ValueError("Serie not found")
        user = user_col.find_one({"_id": user_id})
        if not user:
            raise ValueError("User not found")
        if not user.get("serie_subcribe") or serie_id not in user.get("serie_subcribe"):
            return {"message": "Bạn chưa đăng ký serie này.", "user": user}
        result = unsubscribe_from_topic(serie.get("serie_sns"), user_email)
        if result.get("pendingConfirmation"):
            return result
        user_col.update_one({"_id": user_id}, {"$pull": {"serie_subcribe": serie_id}, "$set": {"updatedAt": None}})
        serie_col.update_one({"_id": serie_col.ObjectId(serie_id)}, {"$inc": {"serie_subcribe_num": -1}, "$set": {"updatedAt": None}})
        return {"message": "Bạn đã hủy đăng ký thành công.", "user": None}
    subs = _SUBSCRIPTIONS.get(serie_id, set())
    if user_id not in subs:
        return {"message": "Bạn chưa đăng ký serie này.", "user": None}
    subs.remove(user_id)
    return {"result": {"serieId": serie_id, "userId": user_id}}


def delete_serie(serie_id):
    db = _db()
    if db:
        serie_col = db.collection("series")
        user_col = db.collection("users")
        from bson import ObjectId

        serie = serie_col.find_one({"_id": ObjectId(serie_id)})
        if not serie:
            raise ValueError("Serie không tồn tại.")
        if serie.get("serie_lessons") and len(serie.get("serie_lessons")) > 0:
            return {"success": False, "warning": "Không thể xóa serie khi vẫn còn bài học trong serie này."}
        user_col.update_many({"serie_subcribe": serie_id}, {"$pull": {"serie_subcribe": serie_id}, "$set": {"updatedAt": None}})
        if serie.get("serie_sns"):
            delete_topic(serie.get("serie_sns"))
        result = serie_col.delete_one({"_id": ObjectId(serie_id)})
        if result.deleted_count > 0 and serie.get("serie_thumbnail"):
            delete_via_cloudfront(serie.get("serie_thumbnail"))
        return result.deleted_count > 0
    return _SERIES.pop(serie_id, None)

File: node_src/services/serie.service.js
const { ObjectId } = require("mongodb");
const { uploadViaCloudFront, deleteViaCloudFront } = require("../utils/s3");
const { v4: uuidv4 } = require("uuid");
const { connectToDatabase } = require("../utils/mongodb");
const {
  createTopic,
  deleteTopic,
  subscribeToSerie,
  unsubscribeFromTopic,
} = require("../utils/sns");
class SerieService {
  async createSerie(data, userId, idToken, file) {
    try {
      //S3 upload
      let imageUrl = "";
      if (file) {
        const uniqueName = `${uuidv4()}_${file.originalname}`;
        imageUrl = await uploadViaCloudFront(
          idToken,
          file.buffer,
          uniqueName,
          file.mimetype,
          `files/user-${userId}/thumbnail`
        );
      }

      const db = await connectToDatabase();
      const serieCollection = db.collection("series");

      const newSerie = {
        ...data,
        serie_thumbnail: imageUrl,
        isPublish: data.isPublish ?? false,
        serie_user: userId,
        serie_lessons: data.serie_lessons ?? [],
        createdAt: new Date(),
        updatedAt: new Date(),
        serie_subcribe_num: 0,
      };

      const result = await serieCollection.insertOne(newSerie);

      const insertedSerieId = result.insertedId.toString();
      //SNS Topic create
      const topicArn = await createTopic(`serie_${insertedSerieId}`);

      // 👉 Cập nhật document series để lưu ARN
      await serieCollection.updateOne(
        { _id: result.insertedId },
        { $set: { serie_sns: topicArn } }
      );

      return { _id: result.insertedId, ...newSerie, serie_sns: topicArn };
    } catch (err) {
      console.error("Error in createSerie:", err);
      throw err;
    }
  }

  async searchSeriesByTitle(keyword) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");

      const query = {
        $text: { $search: keyword },
        isPublish: true,
      };

      // Tùy chọn: sắp xếp theo độ phù hợp
      const projection = {
        score: { $meta: "textScore" },
      };

      const results = await serieCollection
        .find(query, { projection })
        .sort({ score: { $meta: "textScore" } }) // Sắp xếp theo độ phù hợp
        .toArray();

      return results;
    } catch (err) {
      console.error("Error in searchSeriesByTitle:", err);
      throw err;
    }
  }

  async getAllLessonsBySerie(seriesId) {
    try {
      const db = await connectToDatabase();
      const lessonCollection = db.collection("lessons");

      const query = { lesson_serie: seriesId }; // Chỉ lọc theo lesson_serie
      return await lessonCollection.find(query).toArray();
    } catch (err) {
      console.error("Error in getAllLessonsBySerie:", err);
      throw err;
    }
  }
  async getAllSeries(query = {}) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");

      if (query._id) {
        query._id = new ObjectId(query._id);
      }
      return await serieCollection.find(query).toArray();
    } catch (err) {
      console.error("Error in getAllSeries:", err);
      throw err;
    }
  }
  async getAllSeriesByUser(userId) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");

      const series = await serieCollection
        .find({ serie_user: userId })
        .toArray();
      return series;
    } catch (err) {
      console.error("Error in getAllSeriesByUser:", err);
      throw err;
    }
  }

  async getSerieById(id) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");

      if (!ObjectId.isValid(id)) {
        throw new Error("Invalid id format");
      }
      return await serieCollection.findOne({ _id: new ObjectId(id) });
    } catch (err) {
      console.error("Error in getSerieById:", err);
      throw err;
    }
  }

  async getSerieSubcribe(userId) {
    try {
      const db = await connectToDatabase();
      const userCollection = db.collection("users");
      const serieCollection = db.collection("series");

      // Lấy mảng ID của các series mà user đã subscribe
      const user = await userCollection.findOne(
        { _id: userId },
        { projection: { serie_subcribe: 1 } }
      );

      if (!user || !user.serie_subcribe || user.serie_subcribe.length === 0) {
        return {
          message: "Bạn chưa đăng ký Serie nào!",
          serieEmpty: true,
        };
      }

      // Convert chuỗi ID sang ObjectId
      const serieIds = user.serie_subcribe.map((id) => new ObjectId(id));

      // Tìm tất cả các series tương ứng với ID
      const subscribedSeries = await serieCollection
        .find({ _id: { $in: serieIds } })
        .toArray();

      return subscribedSeries;
    } catch (error) {
      throw new Error(`Error getting subscribe list: ${error.message}`);
    }
  }
  async updateSerie(id, data, userId, idToken, file) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");
      if (typeof data.isPublish === "string") {
        data.isPublish = data.isPublish === "true";
      }
      data.updatedAt = new Date();

      const currentSerie = await serieCollection.findOne({
        _id: new ObjectId(id),
      });
      if (!currentSerie) {
        return null;
      }

      if (file) {
        // Xóa thumbnail cũ nếu có
        if (currentSerie.serie_thumbnail) {
          await deleteViaCloudFront(currentSerie.serie_thumbnail);
        }

        // Upload file mới lên S3
        const uniqueName = `${uuidv4()}_${file.originalname}`;
        const newImageUrl = await uploadViaCloudFront(
          idToken,
          file.buffer,
          uniqueName,
          file.mimetype,
          `files/user-${userId}/thumbnail`
        );

        // Gán lại giá trị thumbnail mới cho data update
        data.serie_thumbnail = newImageUrl;
      }

      const updateResult = await serieCollection.updateOne(
        { _id: new ObjectId(id) },
        { $set: data }
      );

      if (updateResult.matchedCount === 0) {
        return null;
      }

      const updatedSerie = await serieCollection.findOne({
        _id: new ObjectId(id),
      });
      return updatedSerie;
    } catch (err) {
      console.error("Error in updateSerie:", err);
      throw err;
    }
  }

  async subscribeSerie(id, userId, userEmail) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");
      const userCollection = db.collection("users");

      const serie = await serieCollection.findOne({
        _id: new ObjectId(id),
      });
      if (!serie || !serie.serie_sns) {
        throw new Error("Serie not found");
      }

      const user = await userCollection.findOne({ _id: userId });
      if (!user) {
        throw new Error("User not found");
      }

      if (user.serie_subcribe && user.serie_subcribe.includes(id)) {
        return {
          message: "Bạn đã đăng ký series này rồi.",
          alreadySubscribed: true,
        };
      } else {
        await subscribeToSerie(serie.serie_sns, userEmail);
      }

      // Add serieId to the user's serie_subcribe array, avoid duplicates, and return updated document
      const userUpdate = await userCollection.findOneAndUpdate(
        { _id: userId },
        {
          $addToSet: { serie_subcribe: id },
          $set: { updatedAt: new Date() },
        },
        { returnDocument: "after" }
      );

      await serieCollection.updateOne(
        { _id: new ObjectId(id) },
        {
          $inc: { serie_subcribe_num: 1 },
          $set: { updatedAt: new Date() },
        }
      );
      return userUpdate;
    } catch (error) {
      throw new Error(`Error subcribing: ${error.message}`);
    }
  }

  async unsubscribeSerie(seriesId, userId, userEmail) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");
      const userCollection = db.collection("users");

      const serie = await serieCollection.findOne({
        _id: new ObjectId(seriesId),
      });

      if (!serie || !serie.serie_sns) {
        throw new Error("Serie not found");
      }

      const user = await userCollection.findOne({ _id: userId });
      if (!user) {
        throw new Error("User not found");
      }

      if (!user.serie_subcribe || !user.serie_subcribe.includes(seriesId)) {
        return {
          message: "Bạn chưa đăng ký serie này.",
          user,
        };
      }

      // Gọi tới SNS để hủy subscription
      const result = await unsubscribeFromTopic(serie.serie_sns, userEmail);

      if (result.pendingConfirmation) {
        return result;
      }
      // Xóa seriesId khỏi user.serie_subcribe
      const userUpdate = await userCollection.findOneAndUpdate(
        { _id: userId },
        {
          $pull: { serie_subcribe: seriesId },
          $set: { updatedAt: new Date() },
        },
        { returnDocument: "after" }
      );

      // Giảm số lượng người đăng ký của series
      await serieCollection.updateOne(
        { _id: new ObjectId(seriesId) },
        {
          $inc: { serie_subcribe_num: -1 },
          $set: { updatedAt: new Date() },
        }
      );

      return {
        message:
          "Bạn đã hủy đăng ký thành công. Từ nay bạn sẽ không nhận thông báo nữa!",
        user: userUpdate.value,
      };
    } catch (error) {
      throw new Error(`Error unsubscribing: ${error.message}`);
    }
  }
  async deleteSerie(id) {
    try {
      const db = await connectToDatabase();
      const serieCollection = db.collection("series");
      const userCollection = db.collection("users");
      // Lấy dữ liệu serie để biết được URL ảnh
      const serie = await serieCollection.findOne({ _id: new ObjectId(id) });
      if (!serie) {
        throw new Error("Serie không tồn tại.");
      }
      if (serie.serie_lessons && serie.serie_lessons.length > 0) {
        return {
          success: false,
          warning: "Không thể xóa serie khi vẫn còn bài học trong serie này.",
        };
      }

      await userCollection.updateMany(
        { serie_subcribe: id }, // tìm tất cả user có chứa id trong mảng
        {
          $pull: { serie_subcribe: id }, // xóa id đó khỏi mảng
          $set: { updatedAt: new Date() },
        }
      );

      if (serie.serie_sns) {
        await deleteTopic(serie.serie_sns);
      }

      const result = await serieCollection.deleteOne({
        _id: new ObjectId(id),
      });
      // Nếu xóa thành công và có ảnh thì xóa ảnh
      if (result.deletedCount > 0 && serie.serie_thumbnail) {
        await deleteViaCloudFront(serie.serie_thumbnail);
      }
      console.log("Delete Sucess");
      return result.deletedCount > 0;
    } catch (err) {
      console.error("Error in deleteSerie:", err);
      throw err;
    }
  }
}

module.exports = new SerieService();
